File: scripts/python/scrape_nba_props.py
"""
scrape_nba_props.py — Scrape NBA player props from RotoWire.

Source: https://www.rotowire.com/betting/nba/player-props.php

The page renders prop tables via JavaScript — one table per stat category
(Points, Rebounds, Assists, Threes, Blocks, Steals, Pts+Reb+Ast, Pts+Reb,
Pts+Ast). Each row has a player, line, sportsbook over/under odds, and
RotoWire's own projection.

We compute confidence as scaled |projection - line| / line, and pickSide as
whichever side the projection lands on.

Output: data/processed/latest-player-props.json (same shape as the existing
file produced by ingest-props.ts, so the homepage `PlayerProps` component
just works.)

Run: python scripts/python/scrape_nba_props.py [--dry-run]
"""

import argparse
import json
import re
import sys
import time
from datetime import datetime, timezone
from pathlib import Path

from scrapling.fetchers import StealthyFetcher

URL = "https://www.rotowire.com/betting/nba/player-props.php"

# Map RotoWire heading text → our market key, label, category
MARKETS = {
    "points":           ("player_points",                 "Points",        "core"),
    "rebounds":         ("player_rebounds",               "Rebounds",      "core"),
    "assists":          ("player_assists",                "Assists",       "core"),
    "threes":           ("player_threes",                 "Threes",        "core"),
    "three pointers":   ("player_threes",                 "Threes",        "core"),
    "blocks":           ("player_blocks",                 "Blocks",        "defense"),
    "steals":           ("player_steals",                 "Steals",        "defense"),
    "turnovers":        ("player_turnovers",              "Turnovers",     "defense"),
    "pts+reb+ast":      ("player_points_rebounds_assists","Pts+Reb+Ast",   "combo"),
    "points + rebounds + assists": ("player_points_rebounds_assists","Pts+Reb+Ast","combo"),
    "pts+reb":          ("player_points_rebounds",        "Pts+Reb",       "combo"),
    "points + rebounds":("player_points_rebounds",        "Pts+Reb",       "combo"),
    "pts+ast":          ("player_points_assists",         "Pts+Ast",       "combo"),
    "points + assists": ("player_points_assists",         "Pts+Ast",       "combo"),
    "reb+ast":          ("player_rebounds_assists",       "Reb+Ast",       "combo"),
    "rebounds + assists":("player_rebounds_assists",      "Reb+Ast",       "combo"),
}


def clean(s: str) -> str:
    return re.sub(r"\s+", " ", (s or "")).strip()


def parse_american(s: str) -> int | None:
    s = clean(s).replace("−", "-")  # unicode minus
    m = re.match(r"^[+-]?\d+$", s)
    if not m:
        return None
    return int(s)


def parse_float(s: str) -> float | None:
    s = clean(s).replace("−", "-")
    try:
        return float(s)
    except ValueError:
        return None


def detect_market_key(heading: str) -> tuple[str, str, str] | None:
    h = heading.lower()
    for key, val in MARKETS.items():
        if key in h:
            return val
    return None


def confidence_from_edge(projection: float | None, line: float | None) -> int:
    if projection is None or line is None or line <= 0:
        return 50
    edge = abs(projection - line) / max(line, 1.0)  # fractional edge
    # Map 0-25% edge to 50-95 confidence
    score = 50 + min(45, edge * 180)
    return int(round(score))


def pick_side(projection: float | None, line: float | None) -> str:
    if projection is None or line is None:
        return "over"
    return "over" if projection > line else "under"


def scrape() -> dict:
    print(f"Fetching {URL} ...", flush=True)
    page = StealthyFetcher.fetch(URL, headless=True, network_idle=True)
    if page.status != 200:
        print(f"  ERROR: HTTP {page.status}", file=sys.stderr)
        return {"available": False, "topProps": [], "props": []}

    props: list[dict] = []

    # RotoWire wraps each market in a <section> or <div> with a heading.
    # We look for any block that contains a player-row table.
    sections = page.css("[class*='prop'], [data-section], section, div.market-block")
    seen_section_keys = set()

    # Scan all tables — for each, find the nearest preceding heading text
    # to determine the market.
    tables = page.css("table")
    for tbl in tables:
        # Find the closest heading text before this table
        heading_text = ""
        prev = tbl.find_previous("h1, h2, h3, h4, h5, .section-title, .heading, [class*='title']")
        if prev:
            heading_text = clean(prev.text)
        else:
            # Fall back: search up the DOM for a heading-y class
            ancestor = tbl.parent
            for _ in range(5):
                if ancestor is None:
                    break
                title = ancestor.css_first("h1, h2, h3, h4, h5, .heading")
                if title:
                    heading_text = clean(title.text)
                    break
                ancestor = ancestor.parent

        market_meta = detect_market_key(heading_text)
        if not market_meta:
            continue
        market, label, category = market_meta

        if market in seen_section_keys:
            continue
        seen_section_keys.add(market)

        # Extract rows
        rows = tbl.css("tr")
        for row in rows[1:]:  # skip header
            cells = row.css("td")
            if len(cells) < 4:
                continue

            # Layout (RotoWire common pattern):
            #  [0] Player (with team/opponent in subtitle)
            #  [1] Projection
            #  [2] Line / Over odds
            #  [3] Under odds
            # Real layout varies — be flexible.
            cell_texts = [clean(c.text) for c in cells]

            # Player name + team + opponent — try first cell
            player_text = cell_texts[0] if cell_texts else ""
            player = player_text.split("(")[0].split("\n")[0].strip()
            team_match = re.search(r"\(([^)]+)\)", player_text)
            team = team_match.group(1) if team_match else None

            # Opponent often in another column or as "vs/at OPP" suffix
            opponent = None
            opp_match = re.search(r"(?:vs|@)\s*([A-Z]{2,3})", player_text, re.I)
            if opp_match:
                opponent = opp_match.group(1).upper()

            # Find projection and line — they're floats in the row
            floats = [parse_float(t) for t in cell_texts]
            floats = [f for f in floats if f is not None]
            projection = None
            line = None
            if len(floats) >= 1:
                # First numeric is usually projection or line
                # Prefer the value that looks line-like (often labeled line)
                line = floats[0] if len(floats) == 1 else floats[1]
                projection = floats[0] if len(floats) >= 2 else None

            # American odds — pick out integers with sign
            ints = [parse_american(t) for t in cell_texts]
            ints = [i for i in ints if i is not None and -1000 < i < 1000]
            over_price = ints[0] if len(ints) >= 1 else None
            under_price = ints[1] if len(ints) >= 2 else None

            if not player or line is None:
                continue

            confidence = confidence_from_edge(projection, line)
            side = pick_side(projection, line)
            signals: list[str] = []
            if projection is not None:
                signals.append(f"RotoWire projection {projection:.1f} vs line {line:.1f}")

            props.append({
                "player": player,
                "team": team,
                "opponent": opponent,
                "market": market,
                "marketLabel": label,
                "category": category,
                "line": line,
                "overPrice": over_price,
                "underPrice": under_price,
                "pickSide": side,
                "confidence": confidence,
                "rationaleSignals": signals,
            })

    # Sort by confidence desc; top 25 broad list, top 5 in topProps
    props.sort(key=lambda p: p["confidence"], reverse=True)
    top_props = props[:5]

    return {
        "generatedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "sport": "NBA",
        "source": "rotowire",
        "available": len(props) > 0,
        "topProps": top_props,
        "props": props[:100],
    }


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--dry-run", action="store_true")
    ap.add_argument("--out", default="data/processed/latest-player-props.json")
    args = ap.parse_args()

    if args.dry_run:
        print("dry run — would scrape", URL)
        return 0

    out = Path(args.out)
    out.parent.mkdir(parents=True, exist_ok=True)

    try:
        data = scrape()
    except Exception as e:
        print(f"FAILED: {e}", file=sys.stderr)
        return 1

    out.write_text(json.dumps(data, indent=2), encoding="utf-8")
    print(f"Wrote {len(data.get('props', []))} props ({len(data.get('topProps', []))} top) → {out}", flush=True)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/scrape-nba-props.ts
// Scrape NBA player props from RotoWire by parsing their inline JSON data.
//
// RotoWire's NBA props page (https://www.rotowire.com/betting/nba/player-props.php)
// embeds prop data inline as `data: [...]` arrays inside per-stat <script> blocks
// (one block per prop: pts, reb, ast, threes, blk, stl, to, pra, pr, pa, ra).
// Each row has the same player keyed by sportsbook, e.g.
//   draftkings_pts, draftkings_ptsOver, draftkings_ptsUnder
//
// We compute:
//   - consensus line  = median of book lines for that player
//   - over/under price = median of book prices on each side
//   - confidence      = inverse of book disagreement (tighter = higher conf)
//   - pickSide        = side with the better no-vig fair price across books
//
// Output: data/processed/latest-player-props.json (existing PlayerProps shape)

import { config } from "dotenv";
config();

import fs from "node:fs";
import path from "node:path";
import { chromium } from "playwright";

const URL = "https://www.rotowire.com/betting/nba/player-props.php";
const OUT = path.resolve(process.cwd(), "data/processed/latest-player-props.json");

type Category = "core" | "defense" | "combo";

const PROPS: Array<{ short: string; market: string; label: string; category: Category }> = [
  { short: "pts",       market: "player_points",                  label: "Points",        category: "core" },
  { short: "reb",       market: "player_rebounds",                label: "Rebounds",      category: "core" },
  { short: "ast",       market: "player_assists",                 label: "Assists",       category: "core" },
  { short: "threes",    market: "player_threes",                  label: "Threes",        category: "core" },
  { short: "blk",       market: "player_blocks",                  label: "Blocks",        category: "defense" },
  { short: "stl",       market: "player_steals",                  label: "Steals",        category: "defense" },
  { short: "turnovers", market: "player_turnovers",               label: "Turnovers",     category: "defense" },
  { short: "ptsrebast", market: "player_points_rebounds_assists", label: "Pts+Reb+Ast",   category: "combo" },
  { short: "ptsreb",    market: "player_points_rebounds",         label: "Pts+Reb",       category: "combo" },
  { short: "ptsast",    market: "player_points_assists",          label: "Pts+Ast",       category: "combo" },
  { short: "rebast",    market: "player_rebounds_assists",        label: "Reb+Ast",       category: "combo" },
  { short: "stlblk",    market: "player_blocks_steals",           label: "Stl+Blk",       category: "combo" },
];

const BOOKS = ["draftkings", "fanduel", "mgm", "betrivers", "caesars", "hardrock", "fanatics", "thescore"];

type RawPlayerRow = {
  name: string;
  team: string;
  opp: string;
  [bookKey: string]: string | null;
};

type PropOut = {
  player: string;
  team: string | null;
  opponent: string | null;
  market: string;
  marketLabel: string;
  category: Category;
  line: number;
  overPrice: number | null;
  underPrice: number | null;
  pickSide: "over" | "under";
  confidence: number;
  rationaleSignals: string[];
};

function median(xs: number[]): number {
  if (xs.length === 0) return 0;
  const s = [...xs].sort((a, b) => a - b);
  const m = Math.floor(s.length / 2);
  return s.length % 2 === 0 ? (s[m - 1] + s[m]) / 2 : s[m];
}

function americanToImplied(p: number): number {
  return p > 0 ? 100 / (p + 100) : -p / (-p + 100);
}

function parseAm(v: string | null | undefined): number | null {
  if (!v || v === "null" || v.trim() === "") return null;
  const n = parseFloat(v.replace("−", "-"));
  return Number.isFinite(n) ? n : null;
}

function parseLine(v: string | null | undefined): number | null {
  if (!v || v === "null" || v.trim() === "") return null;
  const n = parseFloat(v);
  return Number.isFinite(n) ? n : null;
}

// Extract the `data: [...]` array out of a script block. Uses a balanced-bracket
// scan rather than regex, since the array can contain nested objects with `]`.
function extractDataArray(scriptText: string): string | null {
  const marker = "data:";
  const idx = scriptText.indexOf(marker);
  if (idx < 0) return null;
  let i = scriptText.indexOf("[", idx);
  if (i < 0) return null;
  const start = i;
  let depth = 0;
  let inStr = false;
  let strChar = "";
  let escaped = false;
  for (; i < scriptText.length; i++) {
    const ch = scriptText[i];
    if (inStr) {
      if (escaped) {
        escaped = false;
        continue;
      }
      if (ch === "\\") {
        escaped = true;
        continue;
      }
      if (ch === strChar) inStr = false;
      continue;
    }
    if (ch === '"' || ch === "'") {
      inStr = true;
      strChar = ch;
      continue;
    }
    if (ch === "[") depth++;
    else if (ch === "]") {
      depth--;
      if (depth === 0) return scriptText.slice(start, i + 1);
    }
  }
  return null;
}

function rowsToProps(
  rows: RawPlayerRow[],
  prop: (typeof PROPS)[number]
): PropOut[] {
  const out: PropOut[] = [];
  for (const row of rows) {
    const lines: number[] = [];
    const overs: number[] = [];
    const unders: number[] = [];
    for (const book of BOOKS) {
      const lineKey = `${book}_${prop.short}`;
      const overKey = `${book}_${prop.short}Over`;
      const underKey = `${book}_${prop.short}Under`;
      const line = parseLine(row[lineKey] as string | null);
      const over = parseAm(row[overKey] as string | null);
      const under = parseAm(row[underKey] as string | null);
      if (line !== null) lines.push(line);
      if (over !== null) overs.push(over);
      if (under !== null) unders.push(under);
    }
    if (lines.length === 0) continue;

    const consensusLine = median(lines);
    const overPrice = overs.length ? Math.round(median(overs)) : null;
    const underPrice = unders.length ? Math.round(median(unders)) : null;

    // pickSide: take whichever has the lower no-vig implied prob (better EV)
    let pickSide: "over" | "under" = "over";
    if (overPrice !== null && underPrice !== null) {
      const overImp = americanToImplied(overPrice);
      const underImp = americanToImplied(underPrice);
      pickSide = overImp <= underImp ? "over" : "under";
    } else if (underPrice !== null && overPrice === null) {
      pickSide = "under";
    }

    // confidence: tighter line spread across books = higher confidence
    const lineSpread = Math.max(...lines) - Math.min(...lines);
    const baseConf = Math.max(0, 100 - lineSpread * 20); // 0.5 line spread => 90; 1.0 => 80
    const confidence = Math.round(Math.min(95, Math.max(35, baseConf)));

    const opp = (row.opp ?? "").toString();
    const cleanOpp = opp.startsWith("@") ? opp.slice(1) : opp;

    out.push({
      player: row.name,
      team: row.team || null,
      opponent: cleanOpp || null,
      market: prop.market,
      marketLabel: prop.label,
      category: prop.category,
      line: consensusLine,
      overPrice,
      underPrice,
      pickSide,
      confidence,
      rationaleSignals: [
        `${lines.length}-book consensus line ${consensusLine} (spread ${lineSpread.toFixed(1)})`,
      ],
    });
  }
  return out;
}

async function scrape(): Promise<{ scripts: Map<string, string> }> {
  console.log(`Fetching ${URL} ...`);
  const browser = await chromium.launch({ headless: true });
  const ctx = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
    viewport: { width: 1440, height: 900 },
  });
  const page = await ctx.newPage();
  await page.goto(URL, { waitUntil: "domcontentloaded" });
  await page.waitForTimeout(2500);

  const scripts = await page.evaluate(() => {
    const out: Array<{ key: string; text: string }> = [];
    for (const s of Array.from(document.querySelectorAll("script"))) {
      const text = s.textContent ?? "";
      const m = text.match(/const\s+prop\s*=\s*"([^"]+)"/);
      if (m && text.includes("loadTableRW")) {
        out.push({ key: m[1], text });
      }
    }
    return out;
  });

  await browser.close();

  const map = new Map<string, string>();
  for (const { key, text } of scripts) map.set(key, text);
  return { scripts: map };
}

async function main() {
  const { scripts } = await scrape();
  console.log(`Found ${scripts.size} prop scripts:`, [...scripts.keys()].join(", "));

  const allProps: PropOut[] = [];
  for (const prop of PROPS) {
    const script = scripts.get(prop.short);
    if (!script) {
      console.log(`  ${prop.short}: skip (no script)`);
      continue;
    }
    const arrayText = extractDataArray(script);
    if (!arrayText) {
      console.log(`  ${prop.short}: skip (no data array)`);
      continue;
    }
    let rows: RawPlayerRow[];
    try {
      rows = JSON.parse(arrayText) as RawPlayerRow[];
    } catch (err) {
      console.log(`  ${prop.short}: parse failed: ${err}`);
      continue;
    }
    const propsForThis = rowsToProps(rows, prop);
    console.log(`  ${prop.short}: ${rows.length} rows → ${propsForThis.length} props`);
    allProps.push(...propsForThis);
  }

  allProps.sort((a, b) => b.confidence - a.confidence);

  const out = {
    generatedAt: new Date().toISOString(),
    sport: "NBA",
    source: "rotowire",
    available: allProps.length > 0,
    topProps: allProps.slice(0, 5),
    props: allProps.slice(0, 200),
  };

  fs.mkdirSync(path.dirname(OUT), { recursive: true });
  fs.writeFileSync(OUT, JSON.stringify(out, null, 2), "utf8");
  console.log(`\n✓ wrote ${out.props.length} props (top ${out.topProps.length}) → ${OUT}`);
}

main().catch(err => {
  console.error("FAILED:", err);
  process.exit(1);
});
